log/sequencer.go

```diff
@@ -288,133 +288,133 @@ func (s logSequencingTask) update(ctx context.Context, leaves []*trillian.LogLea
-		return 0, err
-	}
-	seqSetNodesLatency.Observe(util.SecondsSince(s.timeSource, stageStart), label)
-	stageStart = s.timeSource.Now()
-
-	// Create the log root ready for signing
-	seqTreeSize.Set(float64(merkleTree.Size()), label)
-	newLogRoot := &trillian.SignedLogRoot{
-		RootHash:       merkleTree.CurrentRoot(),
-		TimestampNanos: s.timeSource.Now().UnixNano(),
-		TreeSize:       merkleTree.Size(),
-		LogId:          currentRoot.LogId,
-		TreeRevision:   newVersion,
-	}
-	sig, err := s.signer.SignLogRoot(newLogRoot)
-	if err != nil {
-		glog.Warningf("%v: signer failed to sign root: %v", logID, err)
-		return 0, err
-	}
-	newLogRoot.Signature = sig
+		// Build objects for the nodes to be updated. Because we deduped via the map
+		// each node can only be created / updated once in each tree revision and
+		// they cannot conflict when we do the storage update.
+		targetNodes, err := s.buildNodesFromNodeMap(nodeMap, newVersion)
+		if err != nil {
+			// Probably an internal error with map building, unexpected.
+			glog.Warningf("%v: Failed to build target nodes in sequencer: %v", logID, err)
+			return err
+		}
 
-	if err := tx.StoreSignedLogRoot(ctx, *newLogRoot); err != nil {
-		glog.Warningf("%v: failed to write updated tree root: %v", logID, err)
-		return 0, err
-	}
-	seqStoreRootLatency.Observe(util.SecondsSince(s.timeSource, stageStart), label)
-	stageStart = s.timeSource.Now()
+		// Now insert or update the nodes affected by the above, at the new tree
+		// version.
+		if err := tx.SetMerkleNodes(ctx, targetNodes); err != nil {
+			glog.Warningf("%v: Sequencer failed to set Merkle nodes: %v", logID, err)
+			return err
+		}
+		seqSetNodesLatency.Observe(util.SecondsSince(s.timeSource, stageStart), label)
+		stageStart = s.timeSource.Now()
+
+		// Create the log root ready for signing
+		seqTreeSize.Set(float64(merkleTree.Size()), label)
+		newLogRoot = &trillian.SignedLogRoot{
+			RootHash:       merkleTree.CurrentRoot(),
+			TimestampNanos: s.timeSource.Now().UnixNano(),
+			TreeSize:       merkleTree.Size(),
+			LogId:          currentRoot.LogId,
+			TreeRevision:   newVersion,
+		}
+		sig, err := s.signer.SignLogRoot(newLogRoot)
+		if err != nil {
+			glog.Warningf("%v: signer failed to sign root: %v", logID, err)
+			return err
+		}
+		newLogRoot.Signature = sig
 
-	// The batch is now fully sequenced and we're done
-	if err := tx.Commit(); err != nil {
+		if err := tx.StoreSignedLogRoot(ctx, *newLogRoot); err != nil {
+			glog.Warningf("%v: failed to write updated tree root: %v", logID, err)
+			return err
+		}
+		seqStoreRootLatency.Observe(util.SecondsSince(s.timeSource, stageStart), label)
+		stageStart = s.timeSource.Now()
+
+		// The batch is now fully sequenced and we're done
+		return tx.Commit()
+	})
+	if err != nil {
 		return 0, err
 	}
-	seqCommitLatency.Observe(util.SecondsSince(s.timeSource, stageStart), label)
+	seqCommitLatency.Observe(util.SecondsSince(s.timeSource, start), label)
```

Comment:
Is this still measuring commit latency?

Reply:
Good question - no, this will measure overall transaction latency, including retries in its current form.
I've updated it.